File: clients/js/src/client.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import { EntroQClient } from "./client";

describe("EntroQClient", () => {
  const baseUrl = "http://localhost:9100";
  let client: EntroQClient;

  beforeEach(() => {
    client = new EntroQClient({ baseUrl });
    vi.stubGlobal("fetch", vi.fn());
  });

  it("should generate a consistent claimant ID", () => {
    const cid = client.getClaimantId();
    expect(cid).toHaveLength(16);
    expect(cid).toMatch(/^[0-9a-f]{16}$/);
    expect(client.getClaimantId()).toBe(cid);
  });

  it("should handle tryClaim correctly", async () => {
    const mockTask = {
      id: "task-1",
      version: 1,
      queue: "q1",
      atMs: "1000",
      claimantId: "cid-1",
      value: { data: "hello" },
      createdMs: "1000",
      modifiedMs: "1000",
      claims: 1,
      attempt: 1,
      err: "",
    };

    vi.mocked(fetch).mockResolvedValueOnce({
      ok: true,
      json: async () => ({ task: mockTask }),
    } as Response);

    const resp = await client.tryClaim(["q1"], 10000);
    expect(resp.task).toEqual(mockTask);

    expect(fetch).toHaveBeenCalledWith(`${baseUrl}/api/v0/claim`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        claimantId: client.getClaimantId(),
        queues: ["q1"],
        durationMs: "10000",
        pollMs: "0",
      }),
    });
  });

  it("should handle modify correctly", async () => {
    const mockResp = {
      inserted: [{ id: "new-task", version: 1, queue: "q1" }],
    };

    vi.mocked(fetch).mockResolvedValueOnce({
      ok: true,
      json: async () => mockResp,
    } as Response);

    const resp = await client.modify({
      inserts: [{ queue: "q1", atMs: "0", value: "val" }],
    });
    expect(resp).toEqual(mockResp);

    expect(fetch).toHaveBeenCalledWith(`${baseUrl}/api/v0/modify`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        claimantId: client.getClaimantId(),
        inserts: [{ queue: "q1", atMs: "0", value: "val" }],
      }),
    });
  });

  it("should throw error on failed request", async () => {
    vi.mocked(fetch).mockResolvedValueOnce({
      ok: false,
      status: 403,
      text: async () => "Forbidden",
    } as Response);

    await expect(client.tryClaim(["q1"])).rejects.toThrow(
      "EntroQ request failed (403): Forbidden"
    );
  });

  it("should handle JSON error messages in failed requests", async () => {
    vi.mocked(fetch).mockResolvedValueOnce({
      ok: false,
      status: 400,
      text: async () => JSON.stringify({ message: "Bad Queue Name" }),
    } as Response);

    await expect(client.tryClaim(["q1"])).rejects.toThrow(
      "EntroQ request failed (400): Bad Queue Name"
    );
  });
});

File: clients/elixir/test/client_test.exs
defmodule EntroQ.ClientTest do
  use ExUnit.Case, async: true

  alias EntroQ.{Client, DependencyError}

  defmodule Transport do
    def request(method, url, headers, body, opts) do
      send(Keyword.fetch!(opts, :test_pid), {:request, method, url, headers, body})
      {:ok, Keyword.fetch!(opts, :response)}
    end
  end

  test "generates and reuses claimant IDs" do
    client = Client.new("http://localhost:9100")

    assert Client.claimant_id(client) =~ ~r/^[0-9a-f]{16}$/
    assert Client.claimant_id(client) == Client.claimant_id(client)
  end

  test "try_claim encodes the JSON request" do
    task = %{
      "id" => "t1",
      "version" => 1,
      "queue" => "q",
      "atMs" => "0",
      "claimantId" => "cid",
      "value" => %{"hello" => "world"},
      "createdMs" => "0",
      "modifiedMs" => "0",
      "claims" => 1,
      "attempt" => 0,
      "err" => ""
    }

    client =
      Client.new("http://localhost:9100/",
        claimant_id: "worker-1",
        transport: Transport,
        transport_opts: [
          test_pid: self(),
          response: %{status: 200, body: Jason.encode!(%{"task" => task})}
        ]
      )

    assert {:ok, claimed} = Client.try_claim(client, ["q"], duration_ms: 10_000)
    assert claimed.id == "t1"
    assert claimed.value == %{"hello" => "world"}

    assert_receive {:request, :post, "http://localhost:9100/api/v0/claim", headers, body}
    assert {"content-type", "application/json"} in headers

    assert Jason.decode!(body) == %{
             "claimantId" => "worker-1",
             "queues" => ["q"],
             "durationMs" => "10000",
             "pollMs" => "0"
           }
  end

  test "parses dependency errors from JSON details" do
    body =
      Jason.encode!(%{
        "message" => "conflict",
        "details" => [
          %{"type" => "DELETE", "id" => %{"id" => "t1", "version" => 2, "queue" => "q"}},
          %{
            "type" => "CLAIM",
            "docId" => %{"namespace" => "ns", "id" => "d1", "version" => 4}
          },
          %{"type" => "DETAIL", "msg" => "stale version"}
        ]
      })

    client =
      Client.new("http://localhost:9100",
        transport: Transport,
        transport_opts: [test_pid: self(), response: %{status: 409, body: body}]
      )

    assert {:error, %DependencyError{} = error} = Client.tasks(client, queue: "q")
    assert error.message == "stale version"
    assert [%EntroQ.TaskID{id: "t1", version: 2, queue: "q"}] = error.deletes
    assert [%EntroQ.DocID{namespace: "ns", id: "d1", version: 4}] = error.doc_claims
  end
end
